cocos/particle/renderer/particle-system-renderer-cpu.ts

```diff
@@ -339,17 +340,20 @@ export default class ParticleSystemRendererCPU extends ParticleSystemRendererBas
     }
 
     private doUpdateScale (pass): void {
+        const nodeScale = this._node_scale;
         switch (this._particleSystem?.scaleSpace) {
         case Space.Local:
-            this._particleSystem?.node.getScale(this._node_scale);
+            this._particleSystem?.node.getScale(nodeScale);
             break;
         case Space.World:
-            this._particleSystem?.node.getWorldScale(this._node_scale);
+            this._particleSystem?.node.getWorldScale(nodeScale);
             break;
         default:
             break;
         }
-        pass.setUniform(this._uScaleHandle, this._node_scale);
+        // NOTE: the `_node_scale` should be a Vec3, but we implement `scale` uniform property as a Vec4,
+        // here we pass a temperate Vec4 object to prevent creating Vec4 object every time we set uniform.
+        pass.setUniform(this._uScaleHandle, _tempNodeScale.set(nodeScale.x, nodeScale.y, nodeScale.z));
```

Comment:
Why not change to use vec3 in uniform too?

Reply:
I've checked the shader implementation, it relies on Vec4's w property

Reply:
Shader will automatically add w value.